tools/pinning/DESIGN.md

```diff
@@ -0,0 +1,124 @@
+much easier for package declarations to be reliably interpreted. The
+introduction of pyproject.toml also allows for the use of build systems other
+than setuptools which becomes relevant in the next section of this doc.
+
+## Our pinning system
+
+### Overview
+
+The files under this directory are used to generate Certbot's pinning files.
+The files under `oldest` are used to generate the constraints file used for our
+"oldest" tests while `normal` is used to generate the constraints used
+everywhere else. `common` includes shared files that are used for both sets of
+pinnings.
+
+Under `normal` and `oldest`, there are two files as of writing this. One is a
+pyproject.toml file for use with [Poetry](https://python-poetry.org/) while
+the other is a script that can be run to regenerate pinnings. The
+pyproject.toml file defines a dummy Python package that depends on everything
+we want to pin. This file largely just depends on our own local packages,
+however, extra dependencies can be declared to further constrain package
+versions or to declare additional dependencies.
+
+The reason we use Poetry is that it is somewhat unique among Python packaging
+tools in that when locking dependencies, it makes a best effort approach to do
+this for all environments rather than just the current environment. This
+includes scanning the dependencies for all wheels and sdists of the package and
+recursively resolving dependencies declared through environment markers that
+are not relevant for the current platform. You can see this in action through
+the inclusion of dependencies like pywin32 which we only have a dependency on
+for Windows.
+
+### Potential problems
+
+As of writing this, I'm aware of two potential problems with this pinning
+system. The first is largely described earlier in the doc which is the problem
+of sdists that use code to dynamically declare its dependencies. It's simply
+not feasible to ensure this arbitrary Python code declares its dependencies in
+the same way across all environments. Luckily, this is a largely a theoretical
+problem and I'm aware of no issues with our current dependencies.
+
+The second problem with this approach is that build dependencies are not
+tracked and pinned. Unfortunately, [this seems to be a largely unsolved problem
+in Python right
+now](https://discuss.python.org/t/pinning-build-dependencies/8363). Our tooling
+ensures that when installing build dependencies when using our pinning files
+that versions from the pinning files are used (see
+https://github.com/certbot/certbot/pull/8443 for more info about that), but I'm
+not aware of any tool that automates the process of tracking and pinning down
+build dependencies. For now, if we find any unpinned build dependencies, we can
+declare a dependency on them in pyproject.toml. If a build dependency isn't
+included in the constraints file, pip will use the latest version available on
+PyPI.
+
+## Possible future improvements
```

Comment:
What bad thing happens if we leave this here?

If you're excited about using GitHub Discussions for this purpose, I'll do this, but I hesitate because:

1. I'm personally not sure we need yet another place to document things.
2. I like that this is being kept with the design document describing what we currently have and the limitations of that system.

Reply:
The bad thing is that I fully expect it to become out of date and we forget to update it ever. This PR is already adding another place to document things, as design documents haven't previouslly lived in `DESIGN.md` files, in general. That being said, it sounds like you're thinking of this as "explaining the current limitations", whereas I read it as "here is our todo list," and the former certainly fits in better. But what happens if/when we actually go to implement one of these? Will we actually update it? Or are we not actually planning to do these? If so, let's just call this limitations. If we are planning to do these, we should make a plan for remembering to update this file, or put this information elsewhere.

Reply:
I see. Any documentation certainly has a risk of becoming out of date and I personally think keeping it next to the code helps reduce that risk, but maybe something like a GitHub issue which we're more used to would be better? Again, if you're excited about using and linking to GitHub Discussions here, that's fine with me though.

I wouldn't say this "limitations" or "explaining the current limitations". I think the previous section titled "Potential problems" does that. These were just ideas I had on how one could go about trying to improve things to address these problems. Unless the Python ecosystem changes in some of the ways I describe in this section or we have significant problems with this design (which I don't think we will), I doubt we'll ever do them. I just thought I'd document some ideas I had here after spending a long time thinking about this problem in case anyone ever works on it again.

With this additional explanation, what do you think? Should I change the title to something like "Alternative designs"? Should I delete the section entirely? Should I move this section somewhere else and if so should I link to it here? Whatever you'd like me to do is probably fine with me.

Reply:
Hm, how about we title it "theoretical improvements" or "theoretical future work" or something and call it a day? I think static explanations of how code works are what belong in the code base, and future work we actively plan to do belongs on github. It sounds like the purpose of this section is to describe work that we could have done but don't actually plan to do, to help explain what this code does not do, in which case that's ok to be in the code base, but let's just make it clear that that's what it is.

Reply:
Works for me.